reana_client/validation/complexity.py

```diff
@@ -0,0 +1,211 @@
+                    tree[name]["scatter_params"] = scheduler["scatter"]["parameters"]
+
+            return tree
+
+        return _parse_stages(self.specification["stages"])
+
+    def populate_parameters(self):
+        """Populate parsed Yadage workflow tree with parameter values."""
+
+        def _parse_params(stage, parent_params):
+            parent_params = parent_params.copy()
+            for param in stage["params"]:
+                if isinstance(param["value"], list):
+                    parent_params[param["key"]] = param["value"]
+                elif isinstance(param["value"], dict):
+                    # Example: input_file: {step: init, output: files}
+                    # In this case `files` values should be taken from
+                    # `parent_params` and saved as `input_file`
+                    output = param["value"].get("output", "")
+                    parent_value = parent_params.get(output, "")
+                    parent_params[param["key"]] = parent_value
+                else:
+                    parent_params[param["key"]] = [param["value"]]
+            return parent_params
+
+        def _parse_stages(stages, parent_params):
+            stages = stages.copy()
+            for stage in stages.keys():
+                stage_value = stages[stage]
+                # Handle params
+                params = _parse_params(stage_value, parent_params)
+                stage_value["params"] = params
+                # Handle nested stages
+                stage_value["stages"] = _parse_stages(stage_value["stages"], params)
+            return stages
+
+        return _parse_stages(self.steps, self.input_params)
+
+    def populate_complexity(self):
+        """Calculate number of jobs and memory needed for the parsed Yadage workflow tree."""
+
+        def _parse_stages(stages):
+            stages = stages.copy()
+            for stage in stages.keys():
+                stage_value = stages[stage]
+                complexity = [(1, stage_value["memory_limit"])]
+
+                # Handle nested stages
+                parsed_stages = _parse_stages(stage_value["stages"])
+                stage_value["stages"] = parsed_stages
+                if parsed_stages:
+                    complexity = self._calculate_complexity(parsed_stages)
+
+                # Handle scatter parameters
+                if stage_value["scatter_params"]:
+                    first_param = stage_value["scatter_params"][0]
+                    param_len = len(stage_value["params"].get(first_param, []))
+                    complexity = [(item[0] * param_len, item[1]) for item in complexity]
+
+                stage_value["complexity"] = complexity
+            return stages
+
+        return _parse_stages(self.steps)
+
+
+class CWLComplexityEstimator(ComplexityEstimatorBase):
+    """REANA CWL workflow complexity estimation."""
+
+    # TODO: implement
```

Comment:
Instead of TODO, please `raise NotImplemented`, or even better please return special value of empty list, which means "cannot detect complexity"... Having the same return type for CWL, Serial, Yadage workflows will allow to start implementing the `reana-server` workflow launcher logic.